src/bin/calc.rs

```diff
@@ -0,0 +1,386 @@
+#![deny(warnings)]
+extern crate coreutils;
+use coreutils::extra::OptionalExt;
+use std::env::args;
+use std::io;
+use std::io::Write;
+//use std::rand::FloatMath::*;
+
+#[derive(Debug, Clone)]
+pub enum Token {
+    Plus,
+    Minus,
+    Divide,
+    Multiply,
+    Exponent,
+    OpenParen,
+    CloseParen,
+    Number(String),
+}
+
+impl Token {
+    pub fn to_str(&self) -> &'static str {
+        match self {
+            &Token::Plus       => "Plus",
+            &Token::Minus      => "Minus",
+            &Token::Divide     => "Divide",
+            &Token::Multiply   => "Multiply",
+            &Token::Exponent   => "Exponent",
+            &Token::OpenParen  => "OpenParen",
+            &Token::CloseParen => "CloseParen",
+            &Token::Number(_)  => "Number",
+        }
+    }
+
+    pub fn to_string(&self) -> String {
+        self.to_str().to_owned()
+    }
+}
+
+#[derive(Debug,  Clone)]
+pub enum ParseError {
+    InvalidNumber(String),
+    UnrecognizedToken(String),
+    UnexpectedToken(String, &'static str),
+    UnexpectedEndOfInput,
+    OtherError(String),
+}
+
+#[derive(Clone,Debug)]
+pub struct IntermediateResult {
+    value: f64,
+    tokens_read: usize,
+}
+
+impl IntermediateResult {
+    fn new(value: f64, tokens_read: usize) -> Self {
+        IntermediateResult {
+            value: value,
+            tokens_read: tokens_read, 
+        }
+    }
+}
+
+pub trait OperatorFunctions {
+    fn is_operator(self) -> bool;
+    fn operator_type(self) -> Token;
+}
+
+impl OperatorFunctions for char {
+    fn is_operator(self) -> bool {
+        self == '+' ||
+        self == '-' ||
+        self == '*' ||
+        self == '/' ||
+        self == '^' ||
+        self == '(' ||
+        self == ')' 
+    }
+
+    fn operator_type(self) -> Token {
+        match self {
+            '+' => Token::Plus,
+            '-' => Token::Minus,
+            '/' => Token::Divide,
+            '*' => Token::Multiply,
+            '^' => Token::Exponent,
+            '(' => Token::OpenParen,
+            ')' => Token::CloseParen,
+            _   => coreutils::extra::fail("Invalid operator", &mut io::stderr())
+        }
+    }
+}
+
+pub fn tokenize(input: &str) -> Result<Vec<Token>, ParseError> {
+    // CONSIDER: any smarter way to guess an initial capacity?
+    let mut tokens: Vec<Token> = Vec::with_capacity(25);
```

Comment:
Again no need for type annotation. Also, input_length is a upperbound to the capacity, since the loop conditionally increment with the bound input_length, while each iteration conditionally push to the vector, thus the needed capacity is <= input_length.